fem/ceed/solvers-qcoarsen.cpp

```diff
@@ -0,0 +1,570 @@
+                          CeedQFunction* qfout, CeedQFunctionContext* context_ptr,
+                          CeedQuadMode fine_qmode, CeedQuadMode coarse_qmode)
+{
+   int ierr;
+   Ceed ceed;
+   ierr = CeedOperatorGetCeed(oper, &ceed); CeedChk(ierr);
+
+   CeedQFunction qfin;
+   ierr = CeedOperatorGetQFunction(oper, &qfin); CeedChk(ierr);
+   CeedInt vlength;
+   ierr = CeedQFunctionGetVectorLength(qfin, &vlength); CeedChk(ierr);
+
+   CeedVector assembledqf;
+   CeedElemRestriction rstr_q;
+   ierr = CeedOperatorLinearAssembleQFunction(
+             oper, &assembledqf, &rstr_q, CEED_REQUEST_IMMEDIATE); CeedChk(ierr);
+
+   CeedBasis qbasisctof; // P_Q
+   CeedBasis fine_basis;
+   ierr = CeedOperatorGetActiveBasis(oper, &fine_basis); CeedChk(ierr);
+   ierr = CeedBasisQCoarsen(fine_basis, qcoarse_basis, &qbasisctof,
+                            qorder_reduction, fine_qmode, coarse_qmode); CeedChk(ierr);
+
+   int ncomp_rstr; // components in ElementRestriction rstr_q
+   ierr = CeedElementRestrictionQCoarsen(rstr_q, qbasisctof, coarse_rstr_q,
+                                         &ncomp_rstr); CeedChk(ierr);
+   CeedInt coarse_vlength = vlength; /// this looks wrong but is probably right
+
+   // coarsen: coarsen the vector assembledqf itself (using basisctof == P_Q)
+   ierr = CeedQFunctionCoarsenAssembledVector(assembledqf, rstr_q, qbasisctof,
+                                              coarse_assembledqf); CeedChk(ierr);
+
+   CeedInt coarse_layout[3];
+   ierr = CeedElemRestrictionGetELayout(*coarse_rstr_q, &coarse_layout);
+   CeedChk(ierr);
+
+   // not entirely sure of the next kernel magic
+   std::string qf_file = GetCeedPath() + "/linear_qf.h";
+   std::string qf = qf_file + ":qcoarsen_linearfunc";
+   ierr = CeedQFunctionCreateInterior(ceed, coarse_vlength, qcoarsen_linearfunc,
+                                      qf.c_str(), qfout); CeedChk(ierr);
+
+   struct LinearQFunctionContext * context =
+      (struct LinearQFunctionContext *) calloc(1,
+                                               sizeof(struct LinearQFunctionContext));
+   context->dim = -1;
+   context->ncomp = ncomp_rstr;
+   for (int i = 0; i < 3; ++i)
+   {
+      context->layout[i] = coarse_layout[i];
+   }
+   CeedQFunctionContext qf_context;
+   ierr = CeedQFunctionContextCreate(ceed, &qf_context); CeedChk(ierr);
+   ierr = CeedQFunctionContextSetData(qf_context, CEED_MEM_HOST, CEED_COPY_VALUES,
+                                      sizeof(*context), context); CeedChk(ierr);
+   ierr = CeedQFunctionSetContext(*qfout, qf_context); CeedChk(ierr);
+   *context_ptr = qf_context;
+   free(context);
+
+   ierr = CeedElemRestrictionDestroy(&rstr_q); CeedChk(ierr);
+   ierr = CeedBasisDestroy(&qbasisctof); CeedChk(ierr);
+   ierr = CeedVectorDestroy(&assembledqf); CeedChk(ierr);
+
+   return 0;
+}
+
+/** @brief given a CeedOperator, use the "assembled" qfunction to create your
```

Comment:
I think you could explain a bit more "what" it does, before explaining "how" it does.